source/common/config/udpa_context_params.cc

```diff
@@ -0,0 +1,91 @@
+#include "common/config/udpa_context_params.h"
+
+#include "common/common/macros.h"
+#include "common/protobuf/utility.h"
+
+namespace Envoy {
+namespace Config {
+
+namespace {
+
+using RenderContextParamCb = std::function<std::string(const envoy::config::core::v3::Node& node)>;
+using NodeContextRenderers = absl::flat_hash_map<std::string, RenderContextParamCb>;
+
+RenderContextParamCb directStringFieldRenderer(const std::string& field) {
+  return [field](const envoy::config::core::v3::Node& node) -> std::string {
+    return MessageUtil::getStringField(node, field);
+  };
+}
+
+RenderContextParamCb localityStringFieldRenderer(const std::string& field) {
+  return [field](const envoy::config::core::v3::Node& node) -> std::string {
+    return MessageUtil::getStringField(node.locality(), field);
+  };
+}
+
+std::string buildSemanticVersionRenderer(const envoy::config::core::v3::Node& node) {
+  const auto& semver = node.user_agent_build_version().version();
+  return fmt::format("{}.{}.{}", semver.major_number(), semver.minor_number(), semver.patch());
+}
+
+const NodeContextRenderers& nodeParamCbs() {
+  CONSTRUCT_ON_FIRST_USE(NodeContextRenderers, {"id", directStringFieldRenderer("id")},
```

Comment:
This seems fragile. Can this be determined via reflection somehow? If not reflection, can there be some type of compile/test check that we cover all fields here if new fields are added?

Reply:
The renderers are not automagic, e.g. the version rendering, so I don't think we can use reflection. We don't actually support all the fields in `Node`, e.g. the extensions, as currently they mostly support a reporting rather than query functionality. Can we leave it as a TODO to add these later as needed?

Reply:
OK that's fine. But then should you document which fields are supported per my other comment about improving the docs?

Reply:
Yep, done.